src/dataset.js

```diff
@@ -61,11 +61,71 @@ export class Dataset {
      */
     pushData(data) {
         checkParamOrThrow(data, 'data', 'Array | Object');
+        const payloadLimitBytes = MAX_PAYLOAD_SIZE_BYTES - 32; // create some leeway
+
+        const dispatch = payload => datasets.putItems({ datasetId: this.datasetId, data: payload });
+        const toJsonArray = array => `[${array.join(',')}]`;
+        const checkAndSerialize = (item, index) => {
+            const payload = JSON.stringify(item);
+            const bytes = Buffer.byteLength(payload);
+            if (bytes > payloadLimitBytes) {
+                const s = typeof index === 'number' ? ` at index ${index} ` : ' ';
+                throw new Error(`Data item${s}is too large (size: ${bytes} bytes, limit: ${MAX_PAYLOAD_SIZE_BYTES} bytes)`);
+            }
+            return { bytes, payload };
+        };
 
-        return datasets.putItems({
-            datasetId: this.datasetId,
-            data,
-        });
+        // Handle singular Objects
+        if (!Array.isArray(data)) {
+            try {
+                const { payload } = checkAndSerialize(data);
+                return dispatch(payload);
+            } catch (err) {
+                return Promise.reject(err);
+            }
+        }
+
+        // Handle Arrays
+        let items;
+        let totalBytes = 0;
+        try {
+            items = data.map((item, index) => {
+                const piece = checkAndSerialize(item, index);
+                totalBytes += piece.bytes;
+                return piece;
+            });
+        } catch (err) {
+            return Promise.reject(err);
+        }
+
+        // Send immediately if not over limit
+        if (totalBytes < payloadLimitBytes) return dispatch(toJsonArray(items.map(i => i.payload)));
+
+        let lastChunkBytes = 0;
```

Comment:
I would also create function in utils to make chunking testable separately

`utils.chunkBySize([{ bytes, paylad }, ...], LIMIT_BYTES)`

that would return an array of chunked payloads

`[[paylad1, paylad2, paylad3], [...], [...], ...`

Then we can also remove part above

```
// Send immediately if not over limit
         if (totalBytes < payloadLimitBytes) return dispatch(toJsonArray(items.map(i => i.payload)));
```

because this part will be simply chunked into an array of size 1, we don't need to handle that separately. Then we don't need `totalBytes` and so the part above might be simplified to

```
items = data.map(checkAndSerialize);
```

Reply:
I don't think we'll ever need this function for anything else, but you're right it would make it easier to test it.

Reply:
I would not document it publicly just extract it to make this easier to read and testabale separately